language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs

```diff
@@ -0,0 +1,90 @@
+-- | Combinators.
+
+{-# LANGUAGE OverloadedStrings #-}
+module Language.PlutusCore.StdLib.Data.Function
+    ( getBuiltinConst
+    , getBuiltinSelf
+    , getBuiltinUnroll
+    , getBuiltinFix
+    ) where
+
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Type
+import           Language.PlutusCore.StdLib.Type
+
+-- | 'const' as a PLC term.
+--
+-- > /\ (A B :: *) -> \(x : A) (y : B) -> x
+getBuiltinConst :: Quote (Term TyName Name ())
+getBuiltinConst = do
+    a <- freshTyName () "a"
+    b <- freshTyName () "b"
+    x <- freshName () "x"
+    y <- freshName () "y"
+    return
+        . TyAbs () a (Type ())
+        . TyAbs () b (Type ())
+        . LamAbs () x (TyVar () a)
+        . LamAbs () y (TyVar () b)
+        $ Var () x
+
+-- | @Self@ as a PLC type.
+--
+-- > \(a :: *) -> fix \(self :: *) -> self -> a
+getBuiltinSelf :: Quote (HoledType TyName ())
+getBuiltinSelf = do
+    a    <- freshTyName () "a"
+    self <- freshTyName () "self"
+    return
+        . HoledType self $ \hole ->
+          TyLam () a (Type ())
+        . hole
+        . TyFun () (TyVar () self)
+        $ TyVar () a
+
+-- | @unroll@ as a PLC term.
+--
+-- > /\(a :: *) -> \(s : self a) -> unwrap s s
+getBuiltinUnroll :: Quote (Term TyName Name ())
+getBuiltinUnroll = do
+    builtinSelf <- getBuiltinSelf
+    a <- freshTyName () "a"
+    s <- freshName () "s"
+    let RecursiveType _ builtinSelfA =
+            holedToRecursive . holedTyApp builtinSelf $ TyVar () a
+    return
+        . TyAbs () a (Type ())
+        . LamAbs () s builtinSelfA
+        . Apply () (Unwrap () $ Var () s)
+        $ Var () s
+
+-- | 'fix' as a PLC term.
+--
+-- > /\(a b :: *) -> \(f : (a -> b) -> a -> b) ->
+-- >    unroll {a -> b} (wrap \(s : self (a -> b)) \(x : a) -> f (unroll {a -> b} s) x)
```

Comment:
Perhaps I'm behind the times again, but I'd have expected `fix` to be something like `(a -> a) -> a`, rather than `(a -> b) -> (a -> b) -> (a -> b)`. Am I missing something?

Reply:
[Yep](https://github.com/input-output-hk/plutus-prototype/tree/master/docs/fomega/z-combinator-benchmarks).

Reply:
Interesting. Worth a note here that this is the Z combinator, or a link to the doc.

Reply:
Done.